### PR TITLE
Improve parsing speed of walkfs plugins

### PR DESCRIPTION
**Pull request by JSCU-CNI**
Inspired by https://github.com/fox-it/dissect.ntfs/pull/35 we profiled and improved the `walkfs` plugin. By removing the unnecessary `FilesystemEntry` -> `TargetPath` -> `FilesystemEntry` conversion we see ~ 2x speed improvement on varying underlying container and filesystem formats.

We also introduced `fsutil.walk_ng` since we missed a function that returns a plain iterator for `FilesystemEntry`s instead of tuple separated files, folders and parent paths.

When refactoring the `walkfs` plugin we also had to refactor the `capabilities` plugin to match it's new record structure.

We currently have not added any extra tests that show the speed improvement we have experienced, however we noticed the https://github.com/fox-it/dissect.target/pull/747 PR, which seems to be using some form of `pytest-benchmark`. That certainly looks promising for the dissect project.

Please let us know if `walkfs_ng` should be renamed to something else, consider it a placeholder for a better function name :)

### CODE REVIEW

**Review by pyrco — CHANGES_REQUESTED** [review]
Very nice find of the unnecessary double conversion!

File: dissect/target/filesystem.py [hunk 1]
```diff
@@ -266,6 +264,19 @@ def walk_ext(
         """
         return self.get(path).walk_ext(topdown, onerror, followlinks)
 
+    def walk_ng(self, path: str) -> Iterator[FilesystemEntry]:
```
[the remark this answers is not in the document]
Comment (Schamper) [reply dissect/target/filesystem.py:267]: `recurse`?

[the remark this answers is not in the document]
Comment (Schamper) [reply dissect/target/filesystem.py:267]: Just `recurse`, leave out the `walk`.

File: dissect/target/helpers/fsutil.py [hunk 2]
```diff
@@ -289,6 +290,18 @@ def walk_ext(path_entry, topdown=True, onerror=None, followlinks=False):
...
+def walk_ng(path_entry: filesystem.FilesystemEntry) -> Iterator[filesystem.FilesystemEntry]:
+    """Recursively walk the given FilesystemEntry, yields FilesystemEntry instances."""
+    if not path_entry.is_dir():
+        return
+
+    for child_entry in path_entry.scandir():
+        yield child_entry
+
+        if child_entry.is_dir() and not child_entry.is_symlink():
+            yield from walk_ng(child_entry)
```
Comment (pyrco) [inline dissect/target/helpers/fsutil.py:302]:
```suggestion
    yield path_entry
    
    if not path_entry.is_dir():
        return

    for child_entry in path_entry.scandir():
    
        if child_entry.is_dir() and not child_entry.is_symlink():
            yield from walk_ng(child_entry)
        else:
            yield child_entry
```

This should return the same results for the walkfs plugin (avoid the duplicate outputs) I think.

**Review by pyrco — CHANGES_REQUESTED** [review]

File: dissect/target/helpers/fsutil.py [hunk 3]
```diff
@@ -291,6 +292,20 @@ def walk_ext(path_entry, topdown=True, onerror=None, followlinks=False):
         yield [path_entry], dirs, files
 
 
+def recurse(path_entry: filesystem.FilesystemEntry) -> Iterator[filesystem.FilesystemEntry]:
+    """Recursively walk the given FilesystemEntry, yields FilesystemEntry instances."""
```
Comment (Schamper) [inline dissect/target/helpers/fsutil.py:296]: Can you properly docstring these references to `FilesystemEntry` too?

**Review by Schamper — CHANGES_REQUESTED** [review]
I have some additional suggestions for the capability changes too:

```diff
diff --git a/dissect/target/plugins/filesystem/unix/capability.py b/dissect/target/plugins/filesystem/unix/capability.py
index 4485994..97095dc 100644
--- a/dissect/target/plugins/filesystem/unix/capability.py
+++ b/dissect/target/plugins/filesystem/unix/capability.py
@@ -1,4 +1,3 @@
-import struct
 from enum import IntEnum
 from io import BytesIO
 from typing import Iterator
@@ -86,8 +85,8 @@ class CapabilityPlugin(Plugin):
         if not self.target.has_function("walkfs"):
             raise UnsupportedPluginError("Need walkfs plugin")
 
-        if self.target.os == "windows":
-            raise UnsupportedPluginError("Walkfs not supported on Windows")
+        if not any(fs.__type__ in ("extfs", "xfs") for fs in self.target.filesystems):
+            raise UnsupportedPluginError("Capability plugin only works on EXT and XFS filesystems")
 
     @export(record=CapabilityRecord)
     def capability_binaries(self) -> Iterator[CapabilityRecord]:
@@ -111,7 +110,7 @@ class CapabilityPlugin(Plugin):
 
             for attr in attrs:
                 try:
-                    parsed_attr = parse_attr(attr, BytesIO(attr.value))
+                    permitted, inheritable, effective, root_id = parse_attr(attr.value)
                 except ValueError as e:
                     self.target.log.warning("Could not parse attributes for entry %s: %s", entry, str(e.value))
                     self.target.log.debug("", exc_info=e)
@@ -119,19 +118,25 @@ class CapabilityPlugin(Plugin):
                 yield CapabilityRecord(
                     ts_mtime=entry.lstat().st_mtime,
                     path=entry.path,
-                    **parsed_attr,
+                    permitted=permitted,
+                    inheritable=inheritable,
+                    effective=effective,
+                    root_id=root_id,
                     _target=self.target,
                 )
 
 
-def parse_attr(attr: object, buf: BytesIO) -> dict:
+def parse_attr(attr: bytes) -> tuple[int, list[str], list[str], bool]:
     """Efficiently parse a Linux xattr capability struct.
 
-    Returns: dictionary of parsed capabilities for the given entry.
+    Returns:
+        A tuple of permitted capability names, inheritable capability names, effective flag and ``root_id``.
     """
+    buf = BytesIO(attr)
 
-    # The struct is small enough we can just use struct
-    magic_etc = struct.unpack("<I", buf.read(4))[0]
+    # The struct is small enough we can just use int.from_bytes
+    magic_etc = int.from_bytes(buf.read(4), "little")
+    effective = magic_etc & VFS_CAP_FLAGS_EFFECTIVE != 0
     cap_revision = magic_etc & VFS_CAP_REVISION_MASK
 
     permitted_caps = []
@@ -153,16 +158,15 @@ def parse_attr(attr: object, buf: BytesIO) -> dict:
     else:
         raise ValueError("Unexpected capability revision '%s'" % cap_revision)
 
-    if data_len != (actual_len := len(attr.value)):
+    if data_len != (actual_len := len(attr)):
         raise ValueError("Unexpected capability length (%s vs %s)", data_len, actual_len)
 
     for _ in range(num_caps):
-        permitted_val, inheritable_val = struct.unpack("<2I", buf.read(8))
-        permitted_caps.append(permitted_val)
-        inheritable_caps.append(inheritable_val)
+        permitted_caps.append(int.from_bytes(buf.read(4), "little"))
+        inheritable_caps.append(int.from_bytes(buf.read(4), "little"))
 
     if cap_revision == VFS_CAP_REVISION_3:
-        root_id = struct.unpack("<I", buf.read(4))[0]
+        root_id = int.from_bytes(buf.read(4), "little")
 
     permitted = []
     inheritable = []
@@ -178,9 +182,4 @@ def parse_attr(attr: object, buf: BytesIO) -> dict:
             if caps[cap_index] & (1 << (capability.value & 31)) != 0:
                 results.append(capability.name)
 
-    return {
-        "root_id": root_id,
-        "permitted": permitted,
-        "inheritable": inheritable,
-        "effective": magic_etc & VFS_CAP_FLAGS_EFFECTIVE != 0,
-    }
+    return permitted, inheritable, effective, root_id
diff --git a/tests/plugins/filesystem/unix/test_capability.py b/tests/plugins/filesystem/unix/test_capability.py
index cf1b8b8..5a4edc2 100644
--- a/tests/plugins/filesystem/unix/test_capability.py
+++ b/tests/plugins/filesystem/unix/test_capability.py
@@ -1,10 +1,11 @@
 from unittest.mock import Mock
 
-from dissect.target.filesystem import VirtualFile
+from dissect.target.filesystem import VirtualFile, VirtualFilesystem
 from dissect.target.plugins.filesystem.unix.capability import CapabilityPlugin
+from dissect.target.target import Target
 
 
-def test_capability_plugin(target_unix, fs_unix):
+def test_capability_plugin(target_unix: Target, fs_unix: VirtualFilesystem) -> None:
     # Some fictional capability values
     xattr1 = Mock()
     xattr1.name = "security.capability"
@@ -33,7 +34,7 @@ def test_capability_plugin(target_unix, fs_unix):
     vfile3.lattr.return_value = [xattr3]
     fs_unix.map_file_entry("/path/to/xattr3/file", vfile3)
 
-    target_unix.add_plugin(CapabilityPlugin)
+    target_unix.add_plugin(CapabilityPlugin, check_compatible=False)
 
     results = list(target_unix.capability_binaries())
     assert len(results) == 3
```

#826 should make this neater in the future too.

File: dissect/target/plugins/filesystem/unix/capability.py [hunk 4]
```diff
@@ -82,88 +82,104 @@ class CapabilityPlugin(Plugin):
...
+            - https://github.com/torvalds/linux/blob/master/include/uapi/linux/capability.h
+        """
+
+        for entry in self.target.fs.recurse("/"):
+            if not entry.is_file() or entry.is_symlink():
+                continue
+
+            try:
+                attrs = [attr for attr in entry.lattr() if attr.name == "security.capability"]
+
```
Comment (Schamper) [inline dissect/target/plugins/filesystem/unix/capability.py:105]:
```suggestion
```

File: dissect/target/plugins/filesystem/unix/capability.py [hunk 5]
```diff
@@ -82,88 +82,104 @@ class CapabilityPlugin(Plugin):
...
-                        _target=self.target,
-                    )
+                    permitted, inheritable, effective, root_id = parse_attr(attr.value)
+                except ValueError as e:
+                    self.target.log.warning("Could not parse attributes for entry %s: %s", entry, str(e.value))
+                    self.target.log.debug("", exc_info=e)
+
+                yield CapabilityRecord(
+                    ts_mtime=entry.lstat().st_mtime,
+                    path=entry.path,
```
Comment (Schamper) [inline dissect/target/plugins/filesystem/unix/capability.py:120]:
```suggestion
                    path=self.target.fs.path(entry.path),
```
I think this is the proper way to pass a path string to a `flow.record` `path` type.

File: dissect/target/plugins/filesystem/unix/capability.py [hunk 6]
```diff
@@ -82,88 +82,104 @@ class CapabilityPlugin(Plugin):
...
+                    permitted=permitted,
+                    inheritable=inheritable,
+                    effective=effective,
+                    root_id=root_id,
+                    _target=self.target,
+                )
+
+
+def parse_attr(attr: bytes) -> tuple[int, list[str], list[str], bool]:
```
Comment (Schamper) [inline dissect/target/plugins/filesystem/unix/capability.py:129]:
```suggestion
def parse_attr(attr: bytes) -> tuple[list[str], list[str], bool, int]:
```
My bad, I changed the order later :smile:.

**Review by Schamper — CHANGES_REQUESTED** [review]

File: dissect/target/filesystem.py [hunk 7]
```diff
@@ -578,10 +589,9 @@ def walk(
         topdown: bool = True,
         onerror: Optional[Callable] = None,
         followlinks: bool = False,
-    ) -> Iterator[str]:
-        """Walk a directory and list its contents as strings.
-
-        It walks across all the files inside the entry recursively.
+    ) -> Iterator[tuple[str, str, str]]:
```
Comment (Schamper) [inline dissect/target/filesystem.py:592]:
```suggestion
    ) -> Iterator[tuple[str, list[str], list[str]]]:
```

File: dissect/target/filesystem.py [hunk 8]
```diff
@@ -228,10 +228,9 @@ def walk(
         topdown: bool = True,
         onerror: Optional[Callable] = None,
         followlinks: bool = False,
-    ) -> Iterator[str]:
-        """Walk a directory pointed to by ``path``, returning the string representation of both files and directories.
-
-        It walks across all the files inside ``path`` recursively.
+    ) -> Iterator[tuple[str, str, str]]:
```
Comment (Schamper) [inline dissect/target/filesystem.py:231]:
```suggestion
    ) -> Iterator[tuple[str, list[str], list[str]]]:
```

**Review by Schamper — CHANGES_REQUESTED** [review]

**Review by Schamper — APPROVED** [review]

**Pull request merged**